feat(platform-io): add strict resource handling for external resources

Makio64 opened the pull request:
Fix #1673 

This fix the error when the image link into a `gltf` dont exist anymore. 

This happen Typically after converting `fbx` or `dae` using `assimp`.

prune have to be call after.

Script to test it : 

```js
import { NodeIO } from '../packages/core/dist/index.modern.js';
import { prune } from '../packages/functions/dist/functions.modern.js';
import { resolve } from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';
import { writeFileSync } from 'fs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

async function testStrictResources() {
    console.log('Testing strict resources functionality...\n');

    const io = new NodeIO();

    // Test 1: Default behavior (strict mode = true)
    console.log('--- Test 1: Default behavior (strict mode) ---');
    try {
        const document = await io.read(resolve(__dirname, '../magic-tower.glb'));
        console.log('✗ Expected error, but file loaded successfully');
    } catch (error) {
        console.log('✓ Expected error in strict mode:', error.message);
    }

    // Test 2: With setStrictResources(false)
    console.log('\n--- Test 2: Non-strict mode ---');
    io.setStrictResources(false);
    try {
        const document = await io.read(resolve(__dirname, '../magic-tower.glb'));
        console.log('✓ File loaded successfully in non-strict mode');
        console.log('Document has', document.getRoot().listTextures().length, 'textures');
        console.log('Warnings were logged for missing external images');
        
        // Apply prune to remove unused objects
        await document.transform(prune());
        console.log('✓ Applied prune transform to remove unused attributes');
        
        // Export the processed GLB
        const outputPath = resolve(__dirname, 'output-magic-tower-processed.glb');
        await io.write(outputPath, document);
        console.log('✓ Exported processed file to:', outputPath);
    } catch (error) {
        console.log('✗ Unexpected error in non-strict mode:', error.message);
        console.error(error);
    }

    // Test 3: Test with NinjaTurtle.glb
    console.log('\n--- Test 3: NinjaTurtle.glb in non-strict mode ---');
    try {
        const document = await io.read(resolve(__dirname, '../NinjaTurtle.glb'));
        console.log('✓ NinjaTurtle.glb loaded successfully in non-strict mode');
        console.log('Document has', document.getRoot().listTextures().length, 'textures');
        console.log('Warnings were logged for missing external images');
        
        // Apply prune to remove unused objects
        await document.transform(prune());
        console.log('✓ Applied prune transform to remove unused attributes');
        
        // Export the processed GLB
        const outputPath = resolve(__dirname, 'output-ninja-turtle-processed.glb');
        await io.write(outputPath, document);
        console.log('✓ Exported processed file to:', outputPath);
    } catch (error) {
        console.log('✗ Unexpected error:', error.message);
        console.error(error);
    }

    // Test 4: Reset to strict mode
    console.log('\n--- Test 4: Reset to strict mode ---');
    io.setStrictResources(true);
    try {
        const document = await io.read(resolve(__dirname, '../NinjaTurtle.glb'));
        console.log('✗ Expected error, but file loaded successfully');
    } catch (error) {
        console.log('✓ Expected error in strict mode:', error.message);
    }
}

testStrictResources().catch(console.error);
```

With the ninja turtle and the tower model i got this output 

```bash 
Testing strict resources functionality...

--- Test 1: Default behavior (strict mode) ---
✓ Expected error in strict mode: ENOENT: no such file or directory, open '/glTF-Transform/2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_RoofTiles.jpg'

--- Test 2: Non-strict mode ---
Failed to load external image at index 2: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Bark.jpg
Failed to load external image at index 1: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Wood.jpg
Failed to load external image at index 0: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_StonseWall.jpg
Failed to load external image at index 7: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Folliage.jpg
Failed to load external image at index 6: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Stones.jpg
Failed to load external image at index 5: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_GrassFloor.jpg
Failed to load external image at index 4: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_PolesRoofBranches.jpg
Failed to load external image at index 0: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_StonseWall.jpg
Failed to load external image at index 3: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_RoofTiles.jpg
Failed to load external image at index 1: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Wood.jpg
Failed to load external image at index 4: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_PolesRoofBranches.jpg
Failed to load external image at index 6: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Stones.jpg
Failed to load external image at index 2: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Bark.jpg
Failed to load external image at index 7: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Folliage.jpg
Failed to load external image at index 5: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_GrassFloor.jpg
Removing texture at index 0 that references failed image 0
Removing texture at index 1 that references failed image 1
Removing texture at index 2 that references failed image 2
Removing texture at index 3 that references failed image 3
Removing texture at index 4 that references failed image 4
Removing texture at index 5 that references failed image 5
Removing texture at index 6 that references failed image 6
Removing texture at index 7 that references failed image 7
Missing optional extension, "KHR_materials_specular".
Missing optional extension, "KHR_materials_volume".
Missing optional extension, "FB_ngon_encoding".
✓ File loaded successfully in non-strict mode
Document has 0 textures
Warnings were logged for missing external images
prune: Removed types... Accessor (22)
✓ Applied prune transform to remove unused attributes
✓ Exported processed file to: /glTF-Transform/test-missing-resources/output-magic-tower-processed.glb

--- Test 3: NinjaTurtle.glb in non-strict mode ---
Failed to load external image at index 0: Z:\3D\Projects\Coops_Projects\MODELS_4_Turbosquid\3D Objects\NinjaTurtles\3D\Ninja_G2_BakeCompleteMap.png
Removing texture at index 0 that references failed image 0
Missing optional extension, "KHR_materials_specular".
Missing optional extension, "KHR_materials_volume".
Missing optional extension, "FB_ngon_encoding".
✓ NinjaTurtle.glb loaded successfully in non-strict mode
Document has 0 textures
Warnings were logged for missing external images
prune: Removed types... Accessor (1)
✓ Applied prune transform to remove unused attributes
✓ Exported processed file to: /glTF-Transform/test-missing-resources/output-ninja-turtle-processed.glb

--- Test 4: Reset to strict mode ---
✓ Expected error in strict mode: ENOENT: no such file or directory, open '/glTF-Transform/Z:\3D\Projects\Coops_Projects\MODELS_4_Turbosquid\3D Objects\NinjaTurtles\3D\Ninja_G2_BakeCompleteMap.png'
```

donmccurdy commented:
Thanks @Makio64! 

Hm, the functionality of `_removeTexturesWithFailedImages` feels a bit out of place for the core library. If I understand right, this is pre-processing the [JSONDocument](https://gltf-transform.dev/modules/core/interfaces/JSONDocument) to remove missing textures, and updating indices for references to the remaining textures?

If that's the result your application needs, then I think it might be best for the pre-processing to happen externally — before you pass the JSON into glTF Transform for parsing. There's no special advantage to doing it within the I/O class, it seems.

What I'd had in mind for the 'strict resources' mode to do would be to leave the missing textures in the resulting document, with their 'image' attributes assigned to 'null'. Then no changes to the JSONDocument should be necessary. User applications might need to deal with the textures somehow (either removing them or supplying the images) before calling operations that like 'textureCompress()' that expect images to be available, but those repairs to the file should be much simpler and safer to do after parsing and constructing a Document.

I'd be happy to merge the PR with that approach, but I'm not sure if that still does what you're hoping it will do?

Makio64 commented:
Thanks for the quick feedback @donmccurdy 

You're right it overstep and added too much code, does it looks good now ? I kept a warning as i think it's important to notify the user about the resources missing.

here the test i used : 

```js
import { NodeIO } from '../packages/core/dist/index.modern.js';
import { prune } from '../packages/functions/dist/functions.modern.js';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

// Count texture references in materials
const countTextureReferences = (materials) => {
    const slots = ['BaseColorTexture', 'NormalTexture', 'OcclusionTexture', 'EmissiveTexture', 'MetallicRoughnessTexture'];
    return materials.reduce((count, material) =>
        count + slots.filter(slot => material[`get${slot}`]()).length, 0);
};

// Process document with null texture removal
const processDocument = async (document, io, outputName) => {
    const root = document.getRoot();
    const textures = root.listTextures();
    const materials = root.listMaterials();

    // Check textures with null images
    const nullTextures = textures.filter(t => t.getImage() === null);
    console.log(`✓ Found ${nullTextures.length}/${textures.length} textures with null images`);

    // Count texture references before
    const refsBefore = countTextureReferences(materials);
    console.log(`✓ Materials have ${refsBefore} texture references`);

    // Detach null textures and prune
    console.log('\n--- Detaching textures with null images ---');
    nullTextures.forEach(t => t.dispose());
    console.log(`✓ Detached ${nullTextures.length} textures`);

    await document.transform(prune());
    console.log('✓ Applied prune transform');

    // Check results
    const remainingTextures = root.listTextures();
    const refsAfter = countTextureReferences(materials);
    console.log(`✓ Textures: ${remainingTextures.length} (was ${textures.length})`);
    console.log(`✓ Texture references: ${refsAfter} (was ${refsBefore})`);

    // Export
    const outputPath = resolve(__dirname, outputName);
    await io.write(outputPath, document);
    console.log(`✓ Exported to: ${outputName}\n`);
};

// Test strict mode behavior
const testStrictMode = async (io, file, shouldFail = true) => {
    try {
        await io.read(resolve(__dirname, file));
        console.log(shouldFail ? '✗ Expected error, but loaded successfully' : '✓ Loaded successfully');
        return true;
    } catch (error) {
        console.log(shouldFail ? `✓ Expected error: ${error.message}` : `✗ Unexpected error: ${error.message}`);
        return false;
    }
};

async function testStrictResources() {
    console.log('Testing strict resources functionality...\n');
    const io = new NodeIO();

    // Test files
    const testFiles = [
        { name: '../magic-tower.glb', output: 'output-magic-tower-processed.glb' },
        { name: '../NinjaTurtle.glb', output: 'output-ninja-turtle-processed.glb' }
    ];

    // Test 1: Strict mode (default)
    console.log('--- Test 1: Strict mode (default) ---');
    await testStrictMode(io, testFiles[0].name);

    // Test 2: Non-strict mode
    console.log('\n--- Test 2: Non-strict mode ---');
    io.setStrictResources(false);

    for (const file of testFiles) {
        console.log(`\n--- Processing ${file.name} ---`);
        const loaded = await testStrictMode(io, file.name, false);

        if (loaded) {
            const document = await io.read(resolve(__dirname, file.name));
            await processDocument(document, io, file.output);
        }
    }

    // Test 3: Reset to strict mode
    console.log('--- Test 3: Reset to strict mode ---');
    io.setStrictResources(true);
    await testStrictMode(io, testFiles[1].name);
}

testStrictResources().catch(console.error);
```

outputing for the test files : 

```bash
Testing strict resources functionality...

--- Test 1: Strict mode (default) ---
✓ Expected error: ENOENT: no such file or directory, open '/Users/davidronai/Desktop/git/makio64-glTF-Transform/2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_StonseWall.jpg'

--- Test 2: Non-strict mode ---

--- Processing ../magic-tower.glb ---
Failed to load external image at index 1: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Wood.jpg
Failed to load external image at index 2: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Bark.jpg
Failed to load external image at index 3: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_RoofTiles.jpg
Failed to load external image at index 4: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_PolesRoofBranches.jpg
Failed to load external image at index 5: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_GrassFloor.jpg
Failed to load external image at index 6: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Stones.jpg
Failed to load external image at index 7: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Folliage.jpg
Failed to load external image at index 0: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_StonseWall.jpg
Failed to load external image at index 1: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Wood.jpg
Failed to load external image at index 2: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Bark.jpg
Failed to load external image at index 3: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_RoofTiles.jpg
Failed to load external image at index 4: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_PolesRoofBranches.jpg
Failed to load external image at index 6: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Stones.jpg
Failed to load external image at index 7: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Folliage.jpg
Failed to load external image at index 5: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_GrassFloor.jpg
Missing optional extension, "KHR_materials_specular".
Missing optional extension, "KHR_materials_volume".
Missing optional extension, "FB_ngon_encoding".
✓ Loaded successfully
Failed to load external image at index 0: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_StonseWall.jpg
Failed to load external image at index 1: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Wood.jpg
Failed to load external image at index 2: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Bark.jpg
Failed to load external image at index 3: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_RoofTiles.jpg
Failed to load external image at index 4: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_PolesRoofBranches.jpg
Failed to load external image at index 5: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_GrassFloor.jpg
Failed to load external image at index 6: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Stones.jpg
Failed to load external image at index 7: 2DAE7_Vander_Meiren_Sander_FinalScene\tex\tex_Folliage.jpg
Missing optional extension, "KHR_materials_specular".
Missing optional extension, "KHR_materials_volume".
Missing optional extension, "FB_ngon_encoding".
✓ Found 8/8 textures with null images
✓ Materials have 10 texture references

--- Detaching textures with null images ---
✓ Detached 8 textures
prune: Removed types... Accessor (22)
✓ Applied prune transform
✓ Textures: 0 (was 8)
✓ Texture references: 0 (was 10)
✓ Exported to: output-magic-tower-processed.glb


--- Processing ../NinjaTurtle.glb ---
Failed to load external image at index 0: Z:\3D\Projects\Coops_Projects\MODELS_4_Turbosquid\3D Objects\NinjaTurtles\3D\Ninja_G2_BakeCompleteMap.png
Missing optional extension, "KHR_materials_specular".
Missing optional extension, "KHR_materials_volume".
Missing optional extension, "FB_ngon_encoding".
✓ Loaded successfully
Failed to load external image at index 0: Z:\3D\Projects\Coops_Projects\MODELS_4_Turbosquid\3D Objects\NinjaTurtles\3D\Ninja_G2_BakeCompleteMap.png
Missing optional extension, "KHR_materials_specular".
Missing optional extension, "KHR_materials_volume".
Missing optional extension, "FB_ngon_encoding".
✓ Found 1/1 textures with null images
✓ Materials have 1 texture references

--- Detaching textures with null images ---
✓ Detached 1 textures
prune: Removed types... Accessor (1)
✓ Applied prune transform
✓ Textures: 0 (was 1)
✓ Texture references: 0 (was 1)
✓ Exported to: output-ninja-turtle-processed.glb

--- Test 3: Reset to strict mode ---
✓ Expected error: ENOENT: no such file or directory, open '/Users/davidronai/Desktop/git/makio64-glTF-Transform/Z:\3D\Projects\Coops_Projects\MODELS_4_Turbosquid\3D Objects\NinjaTurtles\3D\Ninja_G2_BakeCompleteMap.png'
```
